src/guardian/duplicates/keys-duplication-checker.service.ts

```diff
@@ -0,0 +1,180 @@
+import { Inject, Injectable, LoggerService } from '@nestjs/common';
+import { SigningKeyEventsCacheService } from 'contracts/signing-key-events-cache';
+import { SigningKeyEvent } from 'contracts/signing-key-events-cache/interfaces/event.interface';
+import { BlockData } from 'guardian/interfaces';
+import { RegistryKey } from 'keys-api/interfaces/RegistryKey';
+import { WINSTON_MODULE_NEST_PROVIDER } from 'nest-winston';
+
+@Injectable()
+export class KeysDuplicationCheckerService {
+  constructor(
+    @Inject(WINSTON_MODULE_NEST_PROVIDER) private logger: LoggerService,
+    private signingKeyEventsCacheService: SigningKeyEventsCacheService,
+  ) {}
+
+  /**
+   * Identifies and returns duplicated keys.
+   *
+   * Duplicates Search Algorithm:
+   * 1. If there are duplicates within one operator, the key with the lowest index is considered the original, and the others are considered duplicates.
+   * 2. If there are duplicates between different operators, check if a deposited key exists in the duplicates list; all others are considered duplicates.
+   * 3. If there is no deposited key, check the SigningKeyAdded events for operators.
+   * 4. Sort events by block number and logIndex. The earliest event is considered the original, and the others are marked as duplicates.
+   *
+   * What if there is no event for the key? This might indicate that the node returned an incorrect answer, the module didn't implement SigningKeyAdded, or there is a mistake in the cache algorithm. Should we throw an error?
+   *
+   * TODO: If there is no event for the key, should it be considered added in block 0? (Should this be done only for the curated module?)
```

Comment:
Based on the current logic, there should be no keys without events
As far as I remember it could be if new events with indexes were used, but these events were there from the beginning.

Reply:
it is more about case of error in storage.